example/gmt/sphere_preprocessing.c

```diff
@@ -0,0 +1,525 @@
+/*
+  This file is part of p4est.
+  p4est is a C library to manage a collection (a forest) of multiple
+  connected adaptive quadtrees or octrees in parallel.
+
+  Copyright (C) 2010 The University of Texas System
+  Additional copyright (C) 2011 individual authors
+  Written by Carsten Burstedde, Lucas C. Wilcox, and Tobin Isaac
+
+  p4est is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2 of the License, or
+  (at your option) any later version.
+
+  p4est is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with p4est; if not, write to the Free Software Foundation, Inc.,
+  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*/
+
+/** \file sphere_preprocessing.c
+ *
+ * Preprocessing for the sphere model. The main function reads in a list of
+ * geodesics and splits them into segments, such that each segment is 
+ * contained entirely in a single cube face.
+ * 
+ * Usage: p4est_sphere_preprocessing <input.csv>
+ * 
+ * Here <input.csv> is a CSV file where each line
+ *    phi1,theta1,phi2,theta2
+ * represents the unique geodesic between endpoints (phi1, theta1) and
+ * (phi2, theta2). We require that the endpoints are not antipodal, as in this
+ * case there is not a unique geodesic between them.
+ * 
+ * Endpoints are given in spherical coordinates. We take the convention
+ * described here:
+ * https://en.wikipedia.org/wiki/Spherical_coordinate_system.
+ * That is, a spherical coordinate is a pair (phi, theta) where:
+ *  0 <= theta <= 180 is the polar angle
+ *  0 <= phi <= 360   is the azimuth
+ */
+
+#include "gmt_models.h"
+
+/* Convert from angular coordinates to corresponding point on cube face */
+static void
+angular_to_cube (const double angular[2], double xyz[3])
+{
+  double              inf_norm;
+  double              phi, theta;
+
+  /* Convert to radians */
+  phi = angular[0] * M_PI / 180.0;
+  theta = angular[1] * M_PI / 180.0;
+
+  xyz[0] = sin (theta) * cos (phi);
+  xyz[1] = sin (theta) * sin (phi);
+  xyz[2] = cos (theta);
+
+  inf_norm = fmax (fabs (xyz[0]), fmax (fabs (xyz[1]), fabs (xyz[2]))) * 2.0;
```

Comment:
`fmax` is not necessarily included in a C library. Would `SC_MAX` be an option, optionally using a temporary variable to avoid duplicate computation? Also, a performance extremist would store the inverse once and use `*=` three times.